sdk/testing/src/main/java/io/opentelemetry/sdk/testing/junit5/OpenTelemetryExtension.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.testing.junit5;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.propagation.DefaultContextPropagators;
+import io.opentelemetry.exporters.inmemory.InMemorySpanExporter;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.trace.TracerSdkProvider;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor;
+import io.opentelemetry.trace.propagation.HttpTraceContext;
+import java.util.List;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * A JUnit5 extension which sets up the {@link OpenTelemetrySdk} for testing, resetting state
+ * between tests.
+ *
+ * <pre>{@code
+ * > class CoolTest {
+ * >   {@literal @}RegisterExtension
+ * >   static final OpenTelemetryExtension otelTesting = OpenTelemetryExtension.create();
+ * >
+ * >   private final Tracer tracer = otelTesting.getOpenTelemetry().getTracer("test");
+ * >
+ * >   {@literal @}Test
+ * >   void test() {
+ * >     tracer.spanBuilder("name").startSpan();
+ * >     assertThat(otelTesting.getSpans()).containsExactly(expected);
+ * >   }
+ * >  }
+ * }</pre>
+ */
+public class OpenTelemetryExtension
+    implements BeforeEachCallback, BeforeAllCallback, AfterAllCallback {
+
+  /**
+   * Returns a {@link OpenTelemetryExtension} with a default SDK initialized with an in-memory span
+   * exporter and W3C trace context propagation.
+   */
+  public static OpenTelemetryExtension create() {
+    InMemorySpanExporter spanExporter = InMemorySpanExporter.create();
+
+    TracerSdkProvider tracerProvider = TracerSdkProvider.builder().build();
+    tracerProvider.addSpanProcessor(SimpleSpanProcessor.builder(spanExporter).build());
+
+    OpenTelemetrySdk openTelemetry =
+        OpenTelemetrySdk.builder()
+            .setPropagators(
+                DefaultContextPropagators.builder()
+                    .addTextMapPropagator(HttpTraceContext.getInstance())
+                    .build())
+            .setTracerProvider(tracerProvider)
+            .build();
+
+    return new OpenTelemetryExtension(openTelemetry, spanExporter);
+  }
+
+  private final OpenTelemetrySdk openTelemetry;
+  private final InMemorySpanExporter spanExporter;
+
+  private OpenTelemetry previousGlobalOpenTelemetry;
+
+  private OpenTelemetryExtension(
+      OpenTelemetrySdk openTelemetry, InMemorySpanExporter spanExporter) {
+    this.openTelemetry = openTelemetry;
+    this.spanExporter = spanExporter;
+  }
+
+  /** Returns the {@link OpenTelemetrySdk} created by this extension. */
+  public OpenTelemetrySdk getOpenTelemetry() {
```

Comment:
Should we return `OpenTelemetry`, rather than SDK, so people don't create tracers using the SDK directly and expect that to be the normal way to operate? We could have a separate method that returns the TracerManagement interface, perhaps?

Reply:
It's a good point - I wasn't sure how important the distinction is in tests so made it this way. I originally had this in the top level package but moved it into the sdk package / artifact name after realizing it's only useful with the SDK. Instead of exposing API and management separately I figured this is simpler but open to alternatives.

Reply:
My only concern is having people develop good habits wrt. using the API vs. using the SDK, so I think I'd prefer having the 2 methods exposed separately, even if it's just for this test utility. We've had enough people try to use the SDK for writing instrumentation that anything we can do to get them to not do that would be good. :)

Reply:
Makes sense, thanks!